core/src/Instance/Authority.cs

```diff
@@ -118,27 +118,41 @@ public static void ValidateAsUri(string authority)
             }
         }
 
-        private static Authority CreateInstance(string authority, bool validateAuthority)
+        internal static AuthorityType GetAuthorityType(string authority)
         {
-            authority = CanonicalizeUri(authority);
-            ValidateAsUri(authority);
             string[] pathSegments = new Uri(authority).AbsolutePath.Substring(1).Split('/');
-            bool isAdfsAuthority = string.Compare(pathSegments[0], "adfs", StringComparison.OrdinalIgnoreCase) == 0;
-            bool isB2CAuthority = string.Compare(pathSegments[0], B2CAuthority.Prefix, StringComparison.OrdinalIgnoreCase) == 0;
 
-            if (isAdfsAuthority)
+            if (string.Compare(pathSegments[0], "adfs", StringComparison.OrdinalIgnoreCase) == 0)
             {
-                throw CoreExceptionFactory.Instance.GetClientException(
-                    CoreErrorCodes.InvalidAuthorityType, 
-                    "ADFS is not a supported authority");
+                return AuthorityType.Adfs;
             }
-
-            if (isB2CAuthority)
+            else if (string.Compare(pathSegments[0], B2CAuthority.Prefix, StringComparison.OrdinalIgnoreCase) == 0)
+            {
+                return AuthorityType.B2C;
+            }
+            else
             {
-                return new B2CAuthority(authority, validateAuthority);
+                return AuthorityType.Aad;
             }
+        }
+
+        private static Authority CreateInstance(string authority, bool validateAuthority)
+        {
+            authority = CanonicalizeUri(authority);
+            ValidateAsUri(authority);
 
-            return new AadAuthority(authority, validateAuthority);
+            switch (GetAuthorityType(authority))
+            {
+                case AuthorityType.Adfs:
+                    throw CoreExceptionFactory.Instance.GetClientException(CoreErrorCodes.InvalidAuthorityType,
+                       "ADFS is not a supported authority");
+
+                case AuthorityType.B2C:
+                    return new B2CAuthority(authority, validateAuthority);
+
+                default:
+                    return new AadAuthority(authority, validateAuthority);
```

Comment:
suggestion: switch the AuthorityType.Aad and throwing an exception in default. has downsides, thus looking forward to your thoughts on this

Reply:
changed